### PR TITLE
[WIP] dnf rewrite

### PR DESCRIPTION
**Pull request by igor-raits**
It will use only API, no external tools and we will get 5x less code.

### CODE REVIEW

**Comment by gregdek** [comment]
@DJMuggs Please look at this PR; if you give it a +1 we will merge.

**Comment by abadger** [comment]
Note: this pr is still a work in progress. Not ready for review yet.

**Comment by DJMuggs** [comment]
Well this plugin was written for Fedora 21 where dnf wasn't the default yet, so all the yum modules were available.
I haven't had i chance to test it on Fedora 22, lot's of other projects happening.

Glad someone took an intrest in it and is looking to improve the module.

For what i've seen so far of the new code it seems to make sense.
I will help test it when it's ready for testing.

**Comment by ToBeReplaced** [comment]
There is work left to do, but I've hit all of my own requirements, so I'm likely to stop here. If I were to keep working on this, I would start by doing a major restructure and adhere to pep8/pylint standards.

At a bare minimum, we need:
- remove/absent support
- group upgrades

It would be nice to also have:
- disablerepo='*' for use with enablerepo=myrepo to specify a specific set of repositories
- specifying inclusion of optional packages for groups

Are there other requirements that can be enumerated? I don't think it's more than a few hours of work to finish everything. It sure would be nice to provision Fedora 22 machines!

File: packaging/os/dnf.py [hunk 1]
```diff
@@ -22,14 +23,11 @@
...
 import dnf
-
-try:
-    from dnf import find_unfinished_transactions, find_ts_remaining
-    from rpmUtils.miscutils import splitFilename
-    transaction_helpers = True
-except:
-    transaction_helpers = False
+import dnf.cli
+import dnf.util
```
Comment (bcoca) [inline packaging/os/dnf.py]: this should really use the exception handling you removed, we don't want modules to end in an exception, it should be captured as it was before.

**Comment by bcoca** [comment]
we should also at least add a message that absent is not implemented yet, in case someone tries to use and (from my reading of the code) gets no useful information back.

**Comment by ToBeReplaced** [comment]
There seems to be some duplication of work with #704, so I'm giving up on this. There's an open PR from a few days ago to this rewrite at https://github.com/ignatenkobrain/ansible-modules-extras/pull/5 that resolves all known issues and rounds out the functionality. Collaborators may do as they please.

**Comment by gregdek** [comment]
@ignatenkobrain is this still a work in progress, or is it reviewable? Should we be pursuing this PR, or #704 from @shaded-enmity?

**Comment by ToBeReplaced** [comment]
As far as I know, this is complete, barring one issue... `module.fail` should read `module.fail_json` here: https://github.com/ignatenkobrain/ansible-modules-extras/blob/wip/dnf/packaging/os/dnf.py#L233 . I have been using this revision in production since early July without issue.

Comments on the other PR imply that this revision does not support check mode, but that is incorrect, as check mode is supported.

**Comment by ToBeReplaced** [comment]
Comment for the maintainers:

This dnf issue is an example of why my team isn't investigating paying for Ansible Tower.

May 9 - Initial issue reported demonstrating the dnf module does not work on Fedora 22.
May 22 - Rewrite begins.
July 7 - I step in and help out on the rewrite, leaving it in a functioning state for my usage.
July 9 - I outline what else is needed and ask if there are any other requirements.
July 10 - A second rewrite is started elsewhere. I complete all known use-cases from July 9.
July 13 - My commit is merged - PR complete
July 21 - On other PR, maintainer does not review this PR close enough to see that it does support check mode. Does not ask for clarification on this PR, so I don't see that there is something blocking.
August 31 (over 45 days later) - Maintainer asks if this is a work in progress.
September 1 - I clarify the state of the PR.
Today, September 9 - Still no one has reviewed this PR.

It has taken four months, and there is no resolution to date. Moreover, I donated a substantial amount of my time to deliver a quality solution to the problem over two months ago, and it still hasn't been reviewed. Fedora 21 will be EOL by December, at which point all supported Fedora systems will be on dnf.

**Comment by abadger** [comment]
@ToBeReplaced I'm sorry that there's been some miscommunication here.  github is good about reporting comments on tickets but not so good about reporting commits in repositories.  In terms of comments, until your comment on September 1, all the comments seemed to point towards this PR being incomplete and not having anyone to finish off the code:

  "There is work left to do, but I've hit all of my own requirements, so I'm likely to stop here"
  "There seems to be some duplication of work with #704, so I'm giving up on this."

It was good to hear on September 1st that you felt that the code here was complete and ready for testing.  Unfortunately, we're also in the midst of getting ansible-2.0 ready to ship.  We don't have as much time to spend reviewing major changes now as we did in May when we started looking for people to work on this.

If @ignatenkobrain or @DJMuggs or @shaded-enmity or a current extras module maintainer would care to review this, it would help to move this along as our policy for extras modules (of which this is one) is to prioritise review of changes that have been reviewed and upvoted by members of the extras community.

We also would very much like for there to be a maintainer of this code for the future, not just an author, making things work well enough and then abandoning it once it is in.  Is that a role you’re willing to fill?  It makes it quicker to review and merge something in knowing that anything you miss in the review can be fixed by a an active maintainer once a user notices it.

**Comment by shaded-enmity** [comment]
@abadger I'm closing my PR to save you some decision making.

This code looks fine, but the check mode needs to be fixed to be a tad more specific than "changed=True" (which doesn't make any sense in check mode).

**Comment by ToBeReplaced** [comment]
@shaded-enmity `changed=True` is only being sent when there would be a change if check mode were turned off. We should be able to also supply the packages that would be changed. I don't frequently use this feature -- perhaps someone could weigh in on what would be reasonable check-mode behavior?

@abadger I don't think I'm the right person to fulfill the maintainer role, given my generally poor attitude, unfamiliarity with Ansible's development cycle, and lack of incentive (hence the "collaborators may do as they please" comment after feature completion). Ultimately I view the work as a donation, and since this a project backed by a for-profit enterprise, I don't feel the obligation to continue.

I fully understand not wanting to move forward with code that doesn't have a maintainer. I try to polish code I submit for public consideration as well as I can, but perhaps continuing with #704 makes more sense in light of your requirements. I had incorrectly assumed that my commits would continue to fall under the purview of the existing module owner.

**Comment by abadger** [comment]
@ToBeReplaced Thanks.  If worst comes to worst we'll try to get the time to review this and (perhaps not as well as someone who has good domain knowledge) do our best to maintain it.. just takes time...

Regarding check-mode, I think we've recently pushed some changes to the yum module that could be a guide to better output: https://github.com/ansible/ansible-modules-core/pull/1946/files#diff-50dbb2224ca8ab1fc7b6b693f9be32faR798

**Comment by abadger** [comment]
Okay, had a chance to do some minimal testing today.  This looks good.  Merging into devel so it gets into 2.0.

Thanks @ignatenkobrain @ToBeReplaced and @shaded-enmity !  

# Merged

Hi!

This has been merged in, and will also be included in the next major release.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!

**Pull request merged**